### PR TITLE
Add Track, Session and Speaker system

### LINKED ISSUE #58: Track, Session and Speaker system
### The new Event aggregate

image

image

### Example

image

### Notes

* Create/edit forms will be converted to a wizard (with multiple steps)
* Add `IsDraft` to event, so we don't publish the event without finishing it.
* "Create event" link opens a page that user can continue to a draft event or create a new one. If no draft available, directly go to the new event page.
* Save all changes to database, so we don't loose anything.

### PR DESCRIPTION
**Pull request by berkansasmaz**
Resolves #58 

and the design work related to the premium organization process was implemented.

### CODE REVIEW

File: src/EventHub.Application.Contracts/EventHubApplicationContractsModule.cs [hunk 1]
```diff
@@ -11,6 +12,9 @@ namespace EventHub
     )]
     public class EventHubApplicationContractsModule : AbpModule
     {
-
+        public override void ConfigureServices(ServiceConfigurationContext context)
+        {
+            LimitedResultRequestDto.MaxMaxResultCount = 30;
```
Comment (hikalkan) [inline src/EventHub.Application.Contracts/EventHubApplicationContractsModule.cs:17]: Why we needed that? You can add a comment to the code for the next developer.

File: src/EventHub.Application/EventHubApplicationAutoMapperProfile.cs [hunk 2]
```diff
@@ -27,10 +27,20 @@ public EventHubApplicationAutoMapperProfile()
                 .Ignore(x => x.OrganizationDisplayName)
                 .Ignore(x => x.IsLiveNow);
             CreateMap<Event, EventDetailDto>()
+                .ForMember(x => x.Tracks, memberOptions => memberOptions.MapFrom(m => m.Tracks))
                 .Ignore(x => x.OrganizationId)
                 .Ignore(x => x.OrganizationName)
                 .Ignore(x => x.OrganizationDisplayName);
 
+            CreateMap<Track, TrackDto>()
+                .ForMember(x => x.Sessions, memberOptions => memberOptions.MapFrom(m => m.Sessions));
```
Comment (hikalkan) [inline src/EventHub.Application/EventHubApplicationAutoMapperProfile.cs:36]: Property names are the same (`Sessions`) in both classes, is it needed to specify?

File: src/EventHub.Application/Events/EventAppService.cs [hunk 3]
```diff
@@ -232,37 +263,130 @@ public async Task<bool> IsEventOwnerAsync(Guid id)
...
+            return @event.UrlCode;
+        }
+
+        [Authorize]
+        public async Task AddTrackAsync(Guid id, AddTractDto input)
+        {
+            var @event = await _eventRepository.GetAsync(id, true);
+            await CheckIfValidOwnerAsync(@event);
+
+            @event.AddTract(
```
Comment (hikalkan) [inline src/EventHub.Application/Events/EventAppService.cs:302]:
```suggestion
            @event.AddTrack(
```

File: src/EventHub.Application/Events/EventAppService.cs [hunk 4]
```diff
@@ -232,37 +263,130 @@ public async Task<bool> IsEventOwnerAsync(Guid id)
...
+            await _eventRepository.UpdateAsync(@event);
+        }
+
+        [Authorize]
+        public async Task DeleteTrackAsync(Guid id, Guid trackId)
+        {
+            var @event = await _eventRepository.GetAsync(id, true);
+            await CheckIfValidOwnerAsync(@event);
+
+            @event.RemoveTrack(trackId);
```
Comment (hikalkan) [inline src/EventHub.Application/Events/EventAppService.cs:330]: `await _eventRepository.UpdateAsync(@event);`

File: src/EventHub.Application/Events/EventAppService.cs [hunk 5]
```diff
@@ -232,37 +263,130 @@ public async Task<bool> IsEventOwnerAsync(Guid id)
...
+            await _eventRepository.UpdateAsync(@event);
+        }
+
+        [Authorize]
+        public async Task DeleteSessionAsync(Guid id, Guid trackId, Guid sessionId)
+        {
+            var @event = await _eventRepository.GetAsync(id, true);
+            await CheckIfValidOwnerAsync(@event);
+
+            @event.RemoveSession(trackId, sessionId);
```
Comment (hikalkan) [inline src/EventHub.Application/Events/EventAppService.cs:381]: await _eventRepository.UpdateAsync(@event);

File: src/EventHub.Application/Events/EventAppService.cs [hunk 6]
```diff
@@ -277,5 +401,60 @@ private async Task SaveCoverImageAsync(Guid id, IRemoteStreamContent streamConte
 
             await _eventBlobContainer.SaveAsync(blobName, streamContent.GetStream(), overrideExisting: true);
         }
+        
+        private async Task<List<Guid>> GetUserIdsByUserNamesAsync(List<string> userNames)
+        {
+            var userQueryable = await _userRepository.GetQueryableAsync();
+
+            var query = userQueryable
+                    .Where(u => userNames.Any(p => p == u.UserName))
```
Comment (hikalkan) [inline src/EventHub.Application/Events/EventAppService.cs:410]: I think Contains is better than Any. However, that code doesn't guarantee that the returned List of Guids has same element count with the List of strings (usernames). What if a username doesn't exists in the database?

File: src/EventHub.Application/Events/EventAppService.cs [hunk 7]
```diff
@@ -277,5 +401,60 @@ private async Task SaveCoverImageAsync(Guid id, IRemoteStreamContent streamConte
...
+        private List<EventInListDto> GetEventInListDtoFromEventAndOrganizationTupleList(List<(Event @event, Organization organization)> items)
+        {
+            var now = Clock.Now;
+            var events = items.Select(
+                i =>
+                {
+                    var dto = ObjectMapper.Map<Event, EventInListDto>(i.@event);
+                    dto.OrganizationName = i.organization.Name;
+                    dto.OrganizationDisplayName = i.organization.DisplayName;
+                    dto.IsLiveNow = now.IsBetween(i.@event.StartTime, i.@event.EndTime);
```
Comment (hikalkan) [inline src/EventHub.Application/Events/EventAppService.cs:451]: Introducing `IsLive(DateTime now)` method to the `Event` class would be good.

File: src/EventHub.Domain/Events/Speaker.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace EventHub.Events
+{
+    public class Speaker : Entity
+    {
+        public Guid SessionId { get; private set; }
+        
+        public Guid UserId { get; private set; }
+
+        public Speaker(Guid sessionId, Guid userId)
```
Comment (hikalkan) [inline src/EventHub.Domain/Events/Speaker.cs:12]: Also add a private constructor for deserialization (while it seems EF Core can handle it, it is better to do that).

File: src/EventHub.EntityFrameworkCore/EntityFrameworkCore/EventHubDbContextModelCreatingExtensions.cs [hunk 9]
```diff
@@ -71,6 +71,45 @@ public static void ConfigureEventHub(this ModelBuilder builder)
...
+            {
+                b.ToTable(EventHubConsts.DbTablePrefix + "EventSpeakers", EventHubConsts.DbSchema);
+
+                b.ConfigureByConvention();
+
+                b.HasKey(x => new {x.SessionId, x.UserId});
+
+                b.HasOne<IdentityUser>().WithMany().HasForeignKey(x => x.UserId).IsRequired();
+
+                b.HasIndex(x => new {x.UserId, x.SessionId});
```
Comment (hikalkan) [inline src/EventHub.EntityFrameworkCore/EntityFrameworkCore/EventHubDbContextModelCreatingExtensions.cs:112]: No need to define index for primary keys.

File: src/EventHub.Application.Contracts/EventHubApplicationContractsModule.cs (hunk 1 above)
Comment (berkansasmaz) [reply src/EventHub.Application.Contracts/EventHubApplicationContractsModule.cs:17]: I did this to protect ourselves. Currently, requests can be made to the API from places such as Postman. I think this is a necessary limitation for users making requests to the API.

Screen Shot 2021-12-24 at 13 25 53

**Pull request merged**